server/src/main/java/org/apache/druid/segment/indexing/ReaderUtils.java

```diff
@@ -0,0 +1,143 @@
+            LOG.debug("Ignoring columns from JSON path [%s] as path expression is invalid", fields.getExpr());
+            continue;
+          }
+          // Remove the $
+          parsedPath = parsedPath.substring(1);
+          // If the first level is a deep scan, then we need all columns
+          if (parsedPath.length() >= 2 && "..".equals(parsedPath.substring(0, 2))) {
+            return fullInputSchema;
+          }
+          Matcher jsonPathMatcher = JSON_PATH_PATTERN.matcher(parsedPath);
+          if (!jsonPathMatcher.find()) {
+            LOG.warn("Failed to parse JSON path for required column from path [%s]", fields.getExpr());
+            return fullInputSchema;
+          }
+          String matchedGroup = jsonPathMatcher.group();
+          Matcher childMatcher = BRACKET_NOTATED_CHILD_PATTERN.matcher(matchedGroup);
+          if (childMatcher.find()) {
+            // Get name of the column from bracket-notated child i.e. ['region']
+            childMatcher.reset();
+            while (childMatcher.find()) {
+              String columnName = childMatcher.group();
+              // Remove the quote around column name
+              fieldsRequired.add(columnName.substring(1, columnName.length() - 1));
+            }
+          } else if ("[*]".equals(matchedGroup)) {
+            // If the first level is a wildcard, then we need all columns
+            return fullInputSchema;
+          } else {
+            // This can happen if it is a filter expression, slice operator, or index / indexes
+            // We just return all columns...
+            return fullInputSchema;
+          }
+        } else {
+          // Others type aren't supported but returning full schema just in case...
+          LOG.warn("Got unexpected JSONPathFieldType [%s]", fields.getType());
+          return fullInputSchema;
+        }
+      }
+      // If useFieldDiscovery is false then we have already determined all the columns we need to read from
+      // (as only explicitly specified fields will be available to use in the other specs)
+      if (!flattenSpec.isUseFieldDiscovery()) {
+        fieldsRequired.retainAll(fullInputSchema);
+        return fieldsRequired;
+      }
+    }
+
+    // Determine any fields we need to read from parquet file that is used in the transformSpec
+    List<Transform> transforms = transformSpec.getTransforms();
+    for (Transform transform : transforms) {
+      fieldsRequired.addAll(transform.getRequiredColumns());
+    }
+
+    // Determine any fields we need to read from parquet file that is used in the dimensionsSpec
+    List<DimensionSchema> dimensionSchema = dimensionsSpec.getDimensions();
+    for (DimensionSchema dim : dimensionSchema) {
+      fieldsRequired.add(dim.getName());
+    }
+
+    // Determine any fields we need to read from parquet file that is used in the metricsSpec
+    for (AggregatorFactory agg : aggregators) {
+      fieldsRequired.addAll(agg.requiredFields());
+    }
```

Comment:
since dimensions and aggregators can reference transforms, might want to exclude those here since the dim/aggregator would be referring to that instead of the source data.

I've also been thinking about about how to rework this, it feels to me like the object flattener should be fed a list of expected inputs computed from timestampSpec/transforms/dimensions/aggregators, which is sort of what you've got going on here, but I think the flattenSpec handling part should just be baked into flattener stuffs and handled when given this list of required inputs.

Note that I don't think this needs to be changed right now, this utility method is fine, just thinking ahead.

Reply:
I have taken account that dimensions and aggregators can reference transforms by the line below,
`fieldsRequired.retainAll(fullInputSchema);`.

fullInputSchema is the columns of the input file. We are only retaining column names that exist in the input file which will remove all the transformed column.

Reply:
Discussed offline and confirm that `fieldsRequired.retainAll(fullInputSchema);` at the end of the function should eliminate all fields that does not exist in the source input files and is sufficient.